File: src/rider/main/java/com/codex/rider/inspectioncopy/RiderInspectionResultsStartupActivity.java
package com.codex.rider.inspectioncopy;

import com.intellij.openapi.actionSystem.DefaultActionGroup;
import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.startup.StartupActivity;
import com.intellij.openapi.ui.SimpleToolWindowPanel;
import com.intellij.openapi.util.Disposer;
import com.intellij.openapi.wm.ToolWindow;
import com.intellij.openapi.wm.ToolWindowManager;
import com.intellij.ui.content.Content;
import com.intellij.ui.content.ContentManager;
import com.intellij.ui.PopupHandler;
import org.jetbrains.annotations.NotNull;

import javax.swing.JComponent;
import javax.swing.JTree;
import java.awt.Component;
import java.awt.Container;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.Set;

/** Adds a copy-only action to Rider's standard Code Issues result panel. */
public final class RiderInspectionResultsStartupActivity implements StartupActivity.DumbAware {
  private static final String PROBLEMS_TOOL_WINDOW_ID = "Problems View";
  private static final String RIDER_RESULT_PANEL =
    "com.jetbrains.rider.inspections.RiderInspectionsResultPanel";

  @Override
  public void runActivity(@NotNull Project project) {
    Installer installer = new Installer(project);
    Disposer.register(project, installer);
    installer.start();
  }

  private static final class Installer implements com.intellij.openapi.Disposable {
    private final Project project;
    private final Set<JComponent> installedPanels =
      Collections.newSetFromMap(new IdentityHashMap<>());
    private javax.swing.Timer timer;

    private Installer(Project project) {
      this.project = project;
    }

    private void start() {
      ApplicationManager.getApplication().invokeLater(() -> {
        if (project.isDisposed()) return;

        timer = new javax.swing.Timer(500, event -> scan());
        timer.setInitialDelay(250);
        timer.start();
        scan();
      });
    }

    private void scan() {
      if (project.isDisposed()) return;

      ToolWindow toolWindow = ToolWindowManager.getInstance(project)
        .getToolWindow(PROBLEMS_TOOL_WINDOW_ID);
      if (toolWindow == null) return;

      ContentManager contentManager = toolWindow.getContentManager();
      for (Content content : contentManager.getContents()) {
        JComponent panel = findRiderResultPanel(content.getComponent());
        if (panel != null && installedPanels.add(panel)) install(panel);
      }
    }

    private void install(JComponent panel) {
      if (!(panel instanceof SimpleToolWindowPanel)) return;

      JTree tree = findTree(panel);
      if (tree == null) return;

      CopyCurrentInspectionResultsAction copyAction =
        new CopyCurrentInspectionResultsAction(panel, tree);
      DefaultActionGroup actionGroup = new DefaultActionGroup(copyAction);
      PopupHandler.installPopupMenu(tree, actionGroup, "Codex.RiderInspectionCopy.Popup");
    }

    private static JComponent findRiderResultPanel(Component component) {
      if (component instanceof JComponent &&
          RIDER_RESULT_PANEL.equals(component.getClass().getName())) {
        return (JComponent) component;
      }
      if (component instanceof Container) {
        for (Component child : ((Container) component).getComponents()) {
          JComponent result = findRiderResultPanel(child);
          if (result != null) return result;
        }
      }
      return null;
    }

    private static JTree findTree(Component component) {
      if (component instanceof JTree) return (JTree) component;
      if (component instanceof Container) {
        for (Component child : ((Container) component).getComponents()) {
          JTree result = findTree(child);
          if (result != null) return result;
        }
      }
      return null;
    }

    @Override
    public void dispose() {
      if (timer != null) {
        timer.stop();
        timer = null;
      }
      installedPanels.clear();
    }
  }
}

File: src/rider/main/kotlin/com/codex/rider/inspectioncopy/RiderInspectionResultsStartupActivity.kt
package com.codex.rider.inspectioncopy

import com.intellij.openapi.actionSystem.DefaultActionGroup
import com.intellij.openapi.application.ApplicationManager
import com.intellij.openapi.project.Project
import com.intellij.openapi.startup.StartupActivity
import com.intellij.openapi.ui.SimpleToolWindowPanel
import com.intellij.openapi.util.Disposer
import com.intellij.openapi.wm.ToolWindowManager
import com.intellij.ui.PopupHandler
import com.intellij.ui.content.ContentManager
import java.awt.Component
import java.awt.Container
import java.util.Collections
import java.util.IdentityHashMap
import javax.swing.JComponent
import javax.swing.JTree
import javax.swing.Timer

/** Adds a copy-only action to Rider's standard Code Issues result panel. */
class RiderInspectionResultsStartupActivity : StartupActivity.DumbAware {
    override fun runActivity(project: Project) {
        val installer = Installer(project)
        Disposer.register(project, installer)
        installer.start()
    }

    private class Installer(private val project: Project) : com.intellij.openapi.Disposable {
        private val installedPanels: MutableSet<JComponent> =
            Collections.newSetFromMap(IdentityHashMap())
        private var timer: Timer? = null

        fun start() {
            ApplicationManager.getApplication().invokeLater {
                if (project.isDisposed) return@invokeLater

                timer = Timer(500) { scan() }.apply {
                    initialDelay = 250
                    start()
                }
                scan()
            }
        }

        private fun scan() {
            if (project.isDisposed) return

            val toolWindow = ToolWindowManager.getInstance(project)
                .getToolWindow(PROBLEMS_TOOL_WINDOW_ID) ?: return
            val contentManager: ContentManager = toolWindow.contentManager

            for (content in contentManager.contents) {
                val panel = findRiderResultPanel(content.component)
                if (panel != null && installedPanels.add(panel)) {
                    install(panel)
                }
            }
        }

        private fun install(panel: JComponent) {
            if (panel !is SimpleToolWindowPanel) return

            val tree = findTree(panel) ?: return
            val copyAction = CopyCurrentInspectionResultsAction(panel, tree)
            val actionGroup = DefaultActionGroup(copyAction)
            PopupHandler.installPopupMenu(tree, actionGroup, "Codex.RiderInspectionCopy.Popup")
        }

        override fun dispose() {
            timer?.stop()
            timer = null
            installedPanels.clear()
        }

        companion object {
            private const val PROBLEMS_TOOL_WINDOW_ID = "Problems View"
            private const val RIDER_RESULT_PANEL =
                "com.jetbrains.rider.inspections.RiderInspectionsResultPanel"

            private fun findRiderResultPanel(component: Component): JComponent? {
                if (component is JComponent && component.javaClass.name == RIDER_RESULT_PANEL) {
                    return component
                }
                if (component is Container) {
                    for (child in component.components) {
                        findRiderResultPanel(child)?.let { return it }
                    }
                }
                return null
            }

            private fun findTree(component: Component): JTree? {
                if (component is JTree) return component
                if (component is Container) {
                    for (child in component.components) {
                        findTree(child)?.let { return it }
                    }
                }
                return null
            }
        }
    }
}
